File: mobile/android/app/src/test/kotlin/com/codeiy/im/SubscriptionTests.kt
package com.codeiy.im

import android.app.Activity
import com.codeiy.im.core.billing.BillingPackageOption
import com.codeiy.im.core.billing.BillingPurchaseResult
import com.codeiy.im.core.billing.BillingService
import com.codeiy.im.feature.subscription.SubscriptionBackend
import com.codeiy.im.feature.subscription.SubscriptionViewModel
import com.codeiy.im.model.BillingInterval
import com.codeiy.im.model.PlanCode
import com.codeiy.im.model.PlanEntitlements
import com.codeiy.im.model.RadarJson
import com.codeiy.im.model.SubscriptionCatalogPlan
import com.codeiy.im.model.SubscriptionManagement
import com.codeiy.im.model.SubscriptionStatus
import com.codeiy.im.model.SubscriptionUsage
import io.mockk.mockk
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.test.setMain
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Before
import org.junit.Test

@OptIn(ExperimentalCoroutinesApi::class)
class SubscriptionTests {
    private val dispatcher = StandardTestDispatcher()

    @Before fun setUp() = Dispatchers.setMain(dispatcher)
    @After fun tearDown() = Dispatchers.resetMain()

    @Test fun annualSubscriptionKeepsMonthlyUsagePeriod() {
        val value = RadarJson.decodeFromString<SubscriptionUsage>("""
          {"planCode":"pro","subscriptionStatus":"active","periodStart":"2026-07-01T00:00:00Z",
          "periodEnd":"2026-08-01T00:00:00Z","entitlements":{"planCode":"pro","combinedGroupLimit":20,
          "piAgentAnalysisMonthlyLimit":1000},"billingInterval":"annual",
          "billingPeriodStart":"2026-05-10T00:00:00Z","billingPeriodEnd":"2027-05-10T00:00:00Z",
          "usagePeriodStart":"2026-07-01T00:00:00Z","usagePeriodEnd":"2026-08-01T00:00:00Z"}
        """.trimIndent())
        assertEquals(BillingInterval.ANNUAL, value.billingInterval)
        assertNotEquals(value.billingPeriodEnd, value.usagePeriodEnd)
    }

    @Test fun purchaseSuccessSynchronizesBackend() = runTest {
        val backend = FakeBackend()
        val billing = FakeBilling(BillingPurchaseResult.Purchased)
        val model = SubscriptionViewModel(USER_ID, backend, billing)
        model.load().join()
        model.purchase(mockk(relaxed = true), model.packageFor(PlanCode.PLUS)!!).join()
        assertEquals(1, backend.syncCount)
        assertEquals(PlanCode.PLUS, model.state.value.usage?.planCode)
        assertEquals(USER_ID, billing.identifiedUser)
    }

    @Test fun purchaseCancellationDoesNotSynchronizeBackend() = runTest {
        val backend = FakeBackend()
        val model = SubscriptionViewModel(USER_ID, backend, FakeBilling(BillingPurchaseResult.Cancelled))
        model.load().join()
        model.purchase(mockk(relaxed = true), model.packageFor(PlanCode.PLUS)!!).join()
        assertEquals(0, backend.syncCount)
        assertEquals("已取消购买，未产生套餐变更。", model.state.value.message)
    }

    @Test fun restoreIdentifiesUserAndSynchronizesBackend() = runTest {
        val backend = FakeBackend()
        val billing = FakeBilling(BillingPurchaseResult.Purchased)
        val model = SubscriptionViewModel(USER_ID, backend, billing)
        model.restore().join()
        assertEquals(USER_ID, billing.identifiedUser)
        assertEquals(1, billing.restoreCount)
        assertEquals(1, backend.syncCount)
    }

    @Test fun unconfiguredBillingCannotPurchaseAnonymously() = runTest {
        val backend = FakeBackend()
        val billing = FakeBilling(BillingPurchaseResult.Purchased, isConfigured = false)
        val model = SubscriptionViewModel(USER_ID, backend, billing)
        model.load().join()
        model.purchase(mockk(relaxed = true), option).join()
        assertEquals(0, billing.purchaseCount)
        assertEquals(0, backend.syncCount)
    }

    companion object {
        const val USER_ID = "703c41d7-9abd-42ae-9778-b543b489fc51"
        val option = BillingPackageOption("plus_monthly", PlanCode.PLUS, BillingInterval.MONTHLY, "$1")
    }
}

private class FakeBilling(private val result: BillingPurchaseResult, override val isConfigured: Boolean = true) : BillingService {
    var identifiedUser: String? = null
    var restoreCount = 0
    var purchaseCount = 0
    override suspend fun identify(userId: String) { identifiedUser = userId }
    override suspend fun clearIdentity() { identifiedUser = null }
    override suspend fun fetchPackages() = listOf(SubscriptionTests.option)
    override suspend fun purchase(activity: Activity, packageId: String): BillingPurchaseResult { purchaseCount++; return result }
    override suspend fun restorePurchases() { restoreCount++ }
}

private class FakeBackend : SubscriptionBackend {
    var syncCount = 0
    override suspend fun usage() = usage(PlanCode.FREE)
    override suspend fun catalog() = listOf(SubscriptionCatalogPlan(PlanCode.PLUS, "Plus", 1, entitlements(PlanCode.PLUS), listOf(BillingInterval.MONTHLY, BillingInterval.ANNUAL), listOf("plus_monthly", "plus_annual")))
    override suspend fun sync(): SubscriptionUsage { syncCount++; return usage(PlanCode.PLUS) }
    override suspend fun management() = SubscriptionManagement(instruction = "", canOpenInCurrentClient = false)
    private fun entitlements(plan: PlanCode) = PlanEntitlements(plan, 1, 1, 2, 10)
    private fun usage(plan: PlanCode) = SubscriptionUsage(plan, if (plan == PlanCode.FREE) SubscriptionStatus.INACTIVE else SubscriptionStatus.ACTIVE,
        entitlements = entitlements(plan), usagePeriodStart = "2026-07-01T00:00:00Z", usagePeriodEnd = "2026-08-01T00:00:00Z")
}

File: mobile/ios/Tests/SubscriptionTests.swift
import XCTest
@testable import OpportunityRadar

final class SubscriptionTests: XCTestCase {
    func testSubscriptionDTOUsesSeparateBillingAndUsagePeriods() throws {
        let json = """
        {"planCode":"pro","subscriptionStatus":"active","periodStart":"2026-07-01T00:00:00Z",
        "periodEnd":"2026-08-01T00:00:00Z","cancelAtPeriodEnd":false,
        "entitlements":{"planCode":"pro","telegramGroupLimit":null,"wecomGroupLimit":null,
        "combinedGroupLimit":20,"piAgentAnalysisMonthlyLimit":1000},"telegramGroupsUsed":2,
        "wecomGroupsUsed":0,"combinedGroupsUsed":2,"aiAnalysesConsumed":3,"aiAnalysesReserved":1,
        "aiAnalysesRemaining":996,"effectiveStore":"app_store","billingInterval":"annual",
        "billingPeriodStart":"2026-05-10T00:00:00Z","billingPeriodEnd":"2027-05-10T00:00:00Z",
        "usagePeriodStart":"2026-07-01T00:00:00Z","usagePeriodEnd":"2026-08-01T00:00:00Z",
        "entitlementExpiresAt":"2027-05-10T00:00:00Z","willRenew":true,"billingIssue":false,
        "multipleActiveSubscriptions":false,"managementUrl":"https://apps.apple.com/account/subscriptions",
        "lastSyncedAt":"2026-07-12T00:00:00Z"}
        """
        let usage = try APIClient.makeDecoder().decode(SubscriptionUsage.self, from: Data(json.utf8))
        XCTAssertEqual(usage.planCode, .pro)
        XCTAssertEqual(usage.billingInterval, .annual)
        XCTAssertNotEqual(usage.billingPeriodEnd, usage.usagePeriodEnd)
    }

    @MainActor
    func testPurchaseSuccessSynchronizesBackend() async {
        let billing = FakeBillingService(result: .purchased)
        let api = FakeSubscriptionAPI()
        let model = SubscriptionViewModel(api: api, billing: billing, userID: UUID())
        await model.load()
        let option = try! XCTUnwrap(model.package(for: .plus))
        await model.purchase(option)
        XCTAssertEqual(api.syncCount, 1)
        XCTAssertEqual(model.usage?.planCode, .plus)
    }

    @MainActor
    func testPurchaseCancellationDoesNotSynchronizeBackend() async {
        let billing = FakeBillingService(result: .cancelled)
        let api = FakeSubscriptionAPI()
        let model = SubscriptionViewModel(api: api, billing: billing, userID: UUID())
        await model.load()
        await model.purchase(try! XCTUnwrap(model.package(for: .plus)))
        XCTAssertEqual(api.syncCount, 0)
        XCTAssertEqual(model.message, "已取消购买，未产生套餐变更。")
    }

    @MainActor
    func testRestoreSynchronizesBackendAndIdentity() async {
        let billing = FakeBillingService(result: .purchased)
        let api = FakeSubscriptionAPI()
        let userID = UUID()
        let model = SubscriptionViewModel(api: api, billing: billing, userID: userID)
        await model.restore()
        XCTAssertEqual(billing.identifiedUserID, userID)
        XCTAssertEqual(billing.restoreCount, 1)
        XCTAssertEqual(api.syncCount, 1)
    }

    @MainActor
    func testUnconfiguredBillingCannotPurchaseAnonymously() async {
        let billing = FakeBillingService(result: .purchased, configured: false)
        let api = FakeSubscriptionAPI()
        let model = SubscriptionViewModel(api: api, billing: billing, userID: UUID())
        await model.load()
        await model.purchase(BillingPackageOption(id: "plus_monthly", planCode: .plus, interval: .monthly, localizedPrice: "$1"))
        XCTAssertEqual(billing.purchaseCount, 0)
        XCTAssertEqual(api.syncCount, 0)
    }
}

@MainActor
private final class FakeBillingService: BillingService {
    let isConfigured: Bool
    var identifiedUserID: UUID?
    var restoreCount = 0
    var purchaseCount = 0
    let result: BillingPurchaseResult

    init(result: BillingPurchaseResult, configured: Bool = true) {
        self.result = result
        self.isConfigured = configured
    }
    func identify(userID: UUID) async throws { identifiedUserID = userID }
    func clearIdentity() async { identifiedUserID = nil }
    func fetchPackages() async throws -> [BillingPackageOption] {
        [BillingPackageOption(id: "plus_monthly", planCode: .plus, interval: .monthly, localizedPrice: "$1")]
    }
    func purchase(packageID: String) async throws -> BillingPurchaseResult { purchaseCount += 1; return result }
    func restorePurchases() async throws { restoreCount += 1 }
}

private final class FakeSubscriptionAPI: SubscriptionAPI, @unchecked Sendable {
    var syncCount = 0
    func subscription() async throws -> SubscriptionUsage { Self.usage(.free) }
    func subscriptionCatalog() async throws -> [SubscriptionCatalogPlan] {
        [SubscriptionCatalogPlan(planCode: .plus, displayName: "Plus", rank: 1,
          entitlements: Self.entitlements(.plus), availableIntervals: [.monthly, .annual],
          revenuecatPackageIdentifiers: ["plus_monthly", "plus_annual"])]
    }
    func syncSubscription() async throws -> SubscriptionUsage { syncCount += 1; return Self.usage(.plus) }
    func subscriptionManagement() async throws -> SubscriptionManagement {
        SubscriptionManagement(store: nil, managementUrl: nil, instruction: "", canOpenInCurrentClient: false)
    }
    static func entitlements(_ plan: PlanCode) -> PlanEntitlements {
        PlanEntitlements(planCode: plan, telegramGroupLimit: 1, wecomGroupLimit: 1,
                         combinedGroupLimit: 2, piAgentAnalysisMonthlyLimit: 10)
    }
    static func usage(_ plan: PlanCode) -> SubscriptionUsage {
        let now = Date()
        return SubscriptionUsage(planCode: plan, subscriptionStatus: plan == .free ? .inactive : .active,
          periodStart: now, periodEnd: now, cancelAtPeriodEnd: false, entitlements: entitlements(plan),
          telegramGroupsUsed: 0, wecomGroupsUsed: 0, combinedGroupsUsed: 0, aiAnalysesConsumed: 0,
          aiAnalysesReserved: 0, aiAnalysesRemaining: 10, effectiveStore: nil, billingInterval: nil,
          billingPeriodStart: nil, billingPeriodEnd: nil, usagePeriodStart: now, usagePeriodEnd: now,
          entitlementExpiresAt: nil, willRenew: false, billingIssue: false,
          multipleActiveSubscriptions: false, managementUrl: nil, lastSyncedAt: nil)
    }
}
